napari/plugins/utils.py

```diff
@@ -0,0 +1,160 @@
+    """Change the call order of hookimplementations for a pluggy HookCaller.
+
+    Pluggy does not allow a built-in way to change the call order after
+    instantiation.  hookimpls are called in last-in-first-out order.
+    This function accepts the desired call order (a list of plugin names, or
+    plugin modules) and reorders the hookcaller accordingly.
+
+    Parameters
+    ----------
+    hook_caller : pluggy.hooks._HookCaller
+        The hook caller to reorder
+    order : list
+        A list of str, hookimpls, or module_or_class, with the desired
+        CALL ORDER of the hook implementations.
+
+    Raises
+    ------
+    ValueError
+        If the 'order' list cannot be interpreted as a list of "plugin_name"
+        or "plugin" (module_or_class)
+    ValueError
+        if 'order' argument has multiple entries for the same hookimpl
+    """
+    if all(isinstance(o, HookImpl) for o in order):
+        attr = None
+    elif all(isinstance(o, str) for o in order):
+        attr = 'plugin_name'
+    elif any(isinstance(o, str) for o in order):
+        raise TypeError(
+            "order list must be either ALL strings, or ALL modules/classes"
+        )
+    else:
+        attr = 'plugin'
+
+    hookimpls = hook_caller.get_hookimpls()
+    if len(order) > len(hookimpls):
+        raise ValueError(
+            f"too many values ({len(order)} > {len(hookimpls)}) in order."
+        )
+    if attr:
+        hookattrs = [getattr(hookimpl, attr) for hookimpl in hookimpls]
+    else:
+        hookattrs = hookimpls
+
+    # find the current position of items specified in `order`
+    indices = []
+    seen: Set[HookOrderType] = set()
+    for i in order:
+        if i in seen:
+            raise ValueError(
+                f"'order' argument had multiple entries for hookimpl: {i}"
+            )
+        seen.add(i)
+        try:
+            indices.append(hookattrs.index(i))
+        except ValueError as e:
+            msg = f"Could not find hookimpl '{i}'."
+            if attr != 'plugin_name':
+                msg += (
+                    " If all items in `order` "
+                    "argument are not strings, they are assumed to be an "
+                    "imported plugin module or class."
+                )
+            raise ValueError(msg) from e
+
+    # make new arrays for _wrappers and _nonwrappers
+    _wrappers = []
+    _nonwraps = []
+    for i in indices:
+        imp = hookimpls[i]
+        methods = _wrappers if imp.hookwrapper else _nonwraps
+        methods.insert(0, imp)
+
+    # remove items that have been pulled, leaving only items that
+    # were not specified in `order` argument
+    for i in sorted(indices, reverse=True):
+        del hookimpls[i]
+
+    if hookimpls:
+        _wrappers = [x for x in hookimpls if x.hookwrapper] + _wrappers
+        _nonwraps = [x for x in hookimpls if not x.hookwrapper] + _nonwraps
+
+    hook_caller._wrappers = _wrappers
+    hook_caller._nonwrappers = _nonwraps
```

Comment:
> @jni said: honestly I find this function extremely hard to parse. From the docstring, I expected something like:
```python
hook_caller.hooks = [_get_hook_implementation(hook_caller, hook_id) for hook_id in order]
```
with a few checks. Can you explain why it is so much more complicated?

Reply:
sorry again for the complexity, I think this was a case of premature optimization.  It's possible I should restrict the range of inputs this function accepts, which would simplify things quite a bit. But let me first explain what's going on here and what my goals were, and you can guide me as to what I should do.

First off, just a point of clarification: the "`hook_caller`" is the thing _after_ `PluginManger.hook` (e.g. in `plugin_manager.hook.napari_get_reader`, `napari_get_reader` is a `_HookCaller` instance).  It has no attribute called `hooks` (where you might expect to find implementations), so you can't just reorder them like that.  It actually has _multiple_ internal lists storing hook implementations.  One is `._nonwrappers`, where you find the "normal" kind of hook implementations that just get called in sequence... the other is `._wrappers`, where you find "[hookwrappers](https://pluggy.readthedocs.io/en/latest/#wrappers)" (implementations that were decorated with `@hookimpl(hookwrapper=True)`).  These are kind of like context managers that get called both before and after the nonwrappers (see pluggy docs for more).  You can get _all_ hook implementations, regardless of type, using `_HookCaller.get_hookimpls()` (which just returns the combined lists... but is not mutable).  This division of implementation types, and the fact that `_wrappers` must be sorted separately from `_nonwrappers`, makes for part of the difficulty.

Second, `HookImpl` instances have no "id" _per se_ (outside of their object id). They have a `plugin_name` attribute, and a `plugin` attribute (the actual module from which they were imported)... and here's where my premature optimization comes in:  I wanted this function to take in an `order` argument that was a list of _either_ `plugin_name`, `plugin` (module), _or_ the actual `HookImpl` instances themselves. (see the `HookOrderType` I declared above).

Furthermore, in my function, `order` needn't be a complete set of hook implementation pointers.  You can call `permute_hook_implementations(hook_caller, ["my_plugin"])` and it will move  the implementation from `my_plugin` to the front of the list (similar to the [`use_plugin` function in `skimage`](https://github.com/scikit-image/scikit-image/blob/8cf26a04cf20ac81c8233907a02bf0850e3961e0/skimage/io/manage_plugins.py#L212))... (while also being conscious of _hookwrappers_).  Or you can make `order` a (partial) list of `HookImpl` instances, and it will move the ones specified to the front of their respective (wrapper/nonwrapper) lists... this is how I use the function in `qt_plugin_list.py`.
(you can look at `test_permute_hook_implementations` for examples of all the cases I was accommodating).

The function _does_ work at the moment for all of these cases (with tests), but if you'd like to restrict the function signature in the interest of clarifying the logic of the body, let me know how you'd design it.

Reply:
OK! So I finally get this. Mostly. =P

So the thing I wasn't grokking is that wrappers and nonwrappers are in different lists, but here the API is unified, and, I want to say, worse?, it also allows you to mix wrappers and nonwrappers in the same order call? Although I don't know because actually I just looked back at the test and it doesn't look like you test reordering the wrappers?

I can't come up with a useful way to use wrappers for our existing plugins, nor how to even write one. I presume the wrapper signature is not the same as the hookspec signature? Or is it?

My suggestion is that we should (a) rename this function `bring_plugins_to_front`, because that more clearly describes what happens with a partial list, and, more controversially, (b) remove all the support for `_wrappers` for now, adding a note to the function about this. I know it's a bit annoying, but I do think the complexity isn't worth it here, not until we have people wanting to reorder wrappers.

An alternative, if you really want to keep support for wrappers in, might be to have arguments `order=` and `wrapper_order=` and use one or the other, maybe through sub-functions. Is there a specific reason why these things should be on the same list? To me it really feels like putting a list of apples and oranges together, that are never together in the model in the first place.

What do you think?

Reply:
> I can't come up with a useful way to use wrappers for our existing plugins, nor how to even write one. I presume the wrapper signature is not the same as the hookspec signature? Or is it?

have you looked at the [pluggy docs](https://pluggy.readthedocs.io/en/latest/#wrappers) yet?  they are explained pretty well there.  They use the same signature as the hookspec, they cannot return results, they can only modify results using a specific API (`get_result()`/`force_result()`) provided by pluggy.  We do not use them yet, and they would likely never be used in a [`firstresult=True`](https://pluggy.readthedocs.io/en/latest/#first-result-only) hook like `napari_get_reader`.


> My suggestion is that we should (a) rename this function `bring_plugins_to_front`, because that more clearly describes what happens with a partial list, and, more controversially, (b) remove all the support for `_wrappers` for now, adding a note to the function about this. I know it's a bit annoying, but I do think the complexity isn't worth it here, not until we have people wanting to reorder wrappers.

since we're moving farther away from something that could be generalized as a pluggy PR, we can do anything we want. I will rename the function and remove support for wrappers until we use them later.